File: src/tixy.rs
use std::f64::consts::TAU;
use num_traits::identities::Zero;

use crate::num;
use crate::scalar;
use crate::clock::Clock;
use crate::cairo_util::create_image_surface;
use crate::cairo_util::write_image_surface;
use crate::cairo_util::get_image_context;

/// Draw a Tixy-style color field with the parameters given. This
/// function assumes the context is in normalized device coordinates.
fn tixy(
  // The scalar-to-scalar function that defines the color field.
  model: &dyn Fscalar,
  // The current time, in seconds.
  time: num,
  // The number of rows to draw.
  rows: usize,
  // The number of columns to draw.
  cols: usize,
  // The maximum size of a dot drawn by the shader.
  max_radius: num,
  // The Cairo drawing context.
  ctx: &cairo::Context,
) {
  // Draw a border.
  set_color_white(&ctx);
  ctx.new_path();
  ctx.set_line_width(2.0/256.0);
  ctx.rectangle(-1.0, -1.0, 2.0, 2.0);
  ctx.stroke().unwrap();
  // Put the contents of the shader a little bit inside the border.
  ctx.save().unwrap(); {
    ctx.scale(0.95, 0.95);
    for row in 0..rows {
      let dy = row as num / rows as num;
      for col in 0..cols {
        let dx = col as num / cols as num;
        // Transform from [0..col, 0..row] space to [-1..1, -1..1]
        // normalized device coordinates.
        let ndc_x = 2.0*dx-1.0+max_radius;
        let ndc_y = 2.0*dy-1.0+max_radius;
        let source = scalar::new(ndc_x, ndc_y);
        let source = scalar::from_polar(
          source.norm(),
          source.arg()+time*TAU,
        );
        let mut target = scalar::zero();
        // The fractal flame algorithm makes me think it might be
        // interesting to make Tixy-style shaders iterative. This
        // would be more meaningful in the context of Markov
        // algorithms and the new language I want to work on.
        let iterations = 100;
        for _ in 0..iterations {
          target = model.eval(source);
        }
        // The standard Tixy app draws white if the value is below
        // zero and red if the value is above zero. Since I use
        // complex numbers, I want the color to depend on the value's
        // phase instead.
        if target.arg() < 0.0 {
          set_color_white(&ctx);
        } else {
          set_color_super_pink(&ctx);
        }
        // Tweak the radius given by the model.
        //let radius = target.norm()*1.0*0.125;
        //let radius = radius.min(max_radius);
        let range = 2.0;
        let length = target.norm().min(range)/range;
        let radius = length*max_radius;
        // For now, just draw a circle.
        ctx.new_path();
        ctx.arc(ndc_x, ndc_y, radius, 0.0, TAU);
        ctx.fill().unwrap();
      }
    }
  } ctx.restore().unwrap();
}

// TODO: think of a better api for drawing
pub fn draw_sketch_2022_06_16() {
  let width = 1024;
  let height = 1024;
  let ww = width as num * 0.5;
  let hh = height as num * 0.5;
  let length = 6.0;
  let framerate = 15;
  let rows = 16;
  let cols = 16;
  let max_radius = 2.0 / rows as num / 2.0;
  let model = ResNet::with_depth(2);
  let mut clock = Clock::new(length, framerate);
  let mut img = create_image_surface(width, height);
  eprintln!(
    "tixy: {}x{}px {}fps {}sec",
    width,
    height,
    framerate,
    length,
  );
  while clock.is_some() {
    {
      let time = clock.t() / length as num;
      let ctx = get_image_context(&img);
      // Clear the display.
      set_color_space_cadet(&ctx);
      ctx.paint().unwrap();
      ctx.save().unwrap(); {
        // Use normalized device coordinates. The center of the
        // display is the origin of the coordinate system, and the
        // edges are a distance of one away.
        ctx.translate(ww, hh);
        ctx.scale(ww, hh);
        ctx.scale(0.95, 0.95);
        // Draw a border.
        ctx.set_source_rgb(1.0, 1.0, 1.0);
        ctx.new_path();
        ctx.set_line_width(2.0/256.0);
        ctx.rectangle(-1.0, -1.0, 2.0, 2.0);
        ctx.stroke().unwrap();
        /* Draw a tixy-style shader in the upper-left quadrant. */
        ctx.save().unwrap(); {
          ctx.translate(-0.5, -0.5);
          ctx.scale(0.5, 0.5);
          tixy(&model, time, rows, cols, max_radius, &ctx);
        } ctx.restore().unwrap();
        ctx.save().unwrap(); {
          ctx.translate(0.5, -0.5);
          ctx.scale(0.5, 0.5);
          ctx.scale(-1.0, 1.0);
          tixy(&model, time, rows, cols, max_radius, &ctx);
        } ctx.restore().unwrap();
        ctx.save().unwrap(); {
          ctx.translate(-0.5, 0.5);
          ctx.scale(0.5, 0.5);
          ctx.scale(1.0, -1.0);
          tixy(&model, time, rows, cols, max_radius, &ctx);
        } ctx.restore().unwrap();
        ctx.save().unwrap(); {
          ctx.translate(0.5, 0.5);
          ctx.scale(0.5, 0.5);
          ctx.scale(-1.0, -1.0);
          tixy(&model, time, rows, cols, max_radius, &ctx);
        } ctx.restore().unwrap();
      } ctx.restore().unwrap();
    }
    let mut stdout = std::io::stdout();
    write_image_surface(&mut img, &mut stdout);
    clock.tick();
  }
}

fn set_color_super_pink(
  cx: &cairo::Context,
) {
  cx.set_source_rgb(0.839, 0.360, 0.678);
}

fn set_color_space_cadet(
  cx: &cairo::Context,
) {
  cx.set_source_rgb(0.160, 0.160, 0.239);
}

fn set_color_white(
  cx: &cairo::Context,
) {
  cx.set_source_rgb(1.0, 1.0, 1.0);
}

fn set_color_black(
  cx: &cairo::Context,
) {
  cx.set_source_rgb(0.0, 0.0, 0.0);
}

/// Return a random complex number, with a magnitude between 0 and 1,
/// and a phase between 0 and tau.
fn rand_scalar() -> scalar {
  let len = rand::random::<num>().sqrt();
  let phi = rand::random::<num>()*TAU;
  return scalar::from_polar(len, phi);
}

/// A function from one scalar to another.
trait Fscalar {
  fn eval(&self, src: scalar) -> scalar;
}

/// A residual network, used to define random scalar fields.
struct ResNet {
  weight: Vec<scalar>,
}

impl ResNet {
  /// Create a residual network the number of layers provided.
  fn with_depth(depth: usize) -> Self {
    let mut weight = Vec::new();
    for _ in 0..depth {
      let linear = rand_scalar();
      let bias = rand_scalar();
      let threshold = rand_scalar();
      weight.push(linear);
      weight.push(bias);
      weight.push(threshold);
    }
    ResNet { weight }
  }
}

impl Fscalar for ResNet {
  fn eval(&self, mut value: scalar) -> scalar {
    let depth = self.weight.len()/3;
    for i in 0..depth {
      let linear = self.weight[3*i+0];
      let bias = self.weight[3*i+1];
      let threshold = self.weight[3*i+2];
      let mut residual = value*linear+bias;
      if residual.norm_sqr() < threshold.norm_sqr() {
        residual = scalar::zero();
      } else {
        residual = scalar::from_polar(
          residual.norm_sqr(),
          residual.arg()+threshold.arg(),
        );
      }
      value += residual;
    }
    return value;
  }
}

File: 0000-sketch/src/sketch/tixy.py
import sys
import math
import numpy as np
import cairo

from sketch.random_resnet import RandomResNet

tau = 2*math.pi

def debug(*args, **kwargs):
  print(*args, file=sys.stderr, **kwargs)

def dot_factor_latent(latent: np.ndarray) -> float:
  max_length = 5
  point = latent[2:4]
  length = np.linalg.norm(point)
  factor = min(length, max_length)/max_length
  return factor

def set_color_space_cadet(ctx: cairo.Context):
  ctx.set_source_rgb(0.160, 0.160, 0.239)

def set_color_super_pink(ctx: cairo.Context):
  ctx.set_source_rgb(0.839, 0.360, 0.678)

def set_color_white(ctx: cairo.Context):
  ctx.set_source_rgb(1, 1, 1)

palette = [
  set_color_super_pink,
  set_color_white,
]
def set_color_latent(
    latent: np.ndarray,
    ctx: cairo.Context,
):
  point = latent[4:]
  index = np.argmax(point)
  set_color = palette[index]
  set_color(ctx)

frequencies = [
  1,
  1/3,
  1/2,
] 
def clock(time: float):
  value = [
    math.sin(time*frequencies[0]*tau),
    math.cos(time*frequencies[0]*tau),
    math.sin(time*frequencies[1]*tau),
    math.cos(time*frequencies[1]*tau),
    math.sin(time*frequencies[2]*tau),
    math.cos(time*frequencies[2]*tau),
  ]
  return np.array(value)

def draw_color_field(
    model,
    time: np.ndarray,
    origin: np.ndarray,
    ctx: cairo.Context,
):
  for row in range(0, rows):
    dy = row/rows
    for col in range(0, cols):
      dx = col/cols
      ndc_x = 2*dx-1+max_radius
      ndc_y = 2*dy-1+max_radius
      position = np.array([ndc_x, ndc_y, 0, 0, 0, 0])
      latent = model.eval(position+time+origin)
      set_color_latent(latent, ctx)
      radius = dot_factor_latent(latent)*max_radius
      ctx.new_path()
      ctx.arc(ndc_x, ndc_y, radius, 0, tau)
      ctx.fill()
  
if __name__ == '__main__':
  # Application parameters.
  dspw = 256
  dsph = 256
  length = 6
  framerate = 15
  frame_count = framerate*length
  dt = 1/framerate
  rows = 16
  cols = 16
  max_radius = 2 / rows / 2
  origin = np.zeros(2+2+2)

  # Application state.
  model = RandomResNet(dim=2+2+2)

  img = cairo.ImageSurface(
    cairo.FORMAT_ARGB32, dspw, dsph)
  ctx = cairo.Context(img)
  frame = 0

  while frame < frame_count:
    time = clock(frame/framerate)
    # Draw the background.
    set_color_space_cadet(ctx)
    ctx.paint()

    # Enter normalized device coordinates.
    ctx.save()
    ctx.translate(dspw/2, dsph/2)
    ctx.scale(dspw/2, dsph/2)

    # Draw a border.
    ctx.scale(0.95, 0.95)
    set_color_white(ctx)
    ctx.new_path()
    ctx.set_line_width(2/256)
    ctx.rectangle(-1, -1, 2, 2)
    ctx.stroke()
    ctx.scale(0.95, 0.95)

    # Draw a color field in the upper left corner of the display.
    ctx.save()
    ctx.translate(-0.5, -0.5)
    ctx.scale(0.5, 0.5)
    draw_color_field(model, time, origin, ctx)
    ctx.restore()

    # Draw a color field in the upper right corner of the display.
    ctx.save()
    ctx.translate(+0.5, -0.5)
    ctx.scale(0.5, 0.5)
    ctx.scale(-1, 1)
    draw_color_field(model, time, origin, ctx)
    ctx.restore()

    # Draw a color field in the lower left corner of the display.
    ctx.save()
    ctx.translate(-0.5, +0.5)
    ctx.scale(0.5, 0.5)
    ctx.scale(1, -1)
    draw_color_field(model, time, origin, ctx)
    ctx.restore()

    # Draw a color field in the lower right corner of the display.
    ctx.save()
    ctx.translate(+0.5, +0.5)
    ctx.scale(0.5, 0.5)
    ctx.scale(-1, -1)
    draw_color_field(model, time, origin, ctx)
    ctx.restore()

    # Exit normalized device coordinates.
    ctx.restore()

    # Write pixel data for this frame to standard output.
    pixels = img.get_data()
    sys.stdout.buffer.write(pixels)

    # Advance time by one frame.
    frame += 1
